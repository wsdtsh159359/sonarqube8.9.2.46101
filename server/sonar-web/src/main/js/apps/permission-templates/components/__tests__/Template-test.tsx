/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import {
  grantTemplatePermissionToGroup,
  grantTemplatePermissionToUser,
  revokeTemplatePermissionFromGroup,
  revokeTemplatePermissionFromUser
} from '../../../../api/permissions';
import Template from '../Template';

jest.mock('../../../../api/permissions', () => ({
  revokeTemplatePermissionFromUser: jest.fn().mockResolvedValue({}),
  grantTemplatePermissionToUser: jest.fn().mockResolvedValue({}),
  grantTemplatePermissionToGroup: jest.fn().mockResolvedValue({}),
  revokeTemplatePermissionFromGroup: jest.fn().mockResolvedValue({}),
  getPermissionTemplateGroups: jest.fn().mockResolvedValue([]),
  getPermissionTemplateUsers: jest.fn().mockResolvedValue([])
}));

it('render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('revoke group permission if granted', async () => {
  const wrapper = shallowRender();
  const group = {
    name: 'foo',
    permissions: ['bar']
  };
  wrapper.setState({
    groups: [group]
  });
  await wrapper.instance().handleToggleGroup(group, 'bar');
  expect(revokeTemplatePermissionFromGroup).toHaveBeenCalledWith({
    groupName: 'foo',
    templateId: '1',
    permission: 'bar'
  });
});

it('grant group permission', async () => {
  const wrapper = shallowRender();
  const group = {
    name: 'foo',
    permissions: []
  };
  wrapper.setState({
    groups: [group]
  });
  await wrapper.instance().handleToggleGroup(group, 'bar');
  expect(grantTemplatePermissionToGroup).toHaveBeenCalledWith({
    groupName: 'foo',
    templateId: '1',
    permission: 'bar'
  });
});

it('revoke user permission if granted', async () => {
  const wrapper = shallowRender();
  const user = {
    login: 'foo',
    name: 'foo',
    permissions: ['bar']
  };
  wrapper.setState({
    users: [user]
  });
  await wrapper.instance().handleToggleUser(user, 'bar');
  expect(revokeTemplatePermissionFromUser).toHaveBeenCalledWith({
    templateId: '1',
    login: 'foo',
    permission: 'bar'
  });
});

it('grant user permission', async () => {
  const wrapper = shallowRender();
  const user = {
    login: 'foo',
    name: 'foo',
    permissions: []
  };
  wrapper.setState({
    users: [user]
  });
  await wrapper.instance().handleToggleUser(user, 'bar');
  expect(grantTemplatePermissionToUser).toHaveBeenCalledWith({
    templateId: '1',
    login: 'foo',
    permission: 'bar'
  });
});

function shallowRender() {
  return shallow<Template>(
    <Template
      refresh={async () => {}}
      template={{
        id: '1',
        createdAt: '2020-01-01',
        name: 'test',
        defaultFor: [],
        permissions: []
      }}
      topQualifiers={[]}
    />
  );
}
