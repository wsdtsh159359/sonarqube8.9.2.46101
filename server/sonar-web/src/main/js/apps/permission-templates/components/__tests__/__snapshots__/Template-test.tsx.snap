// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  className="page page-limited"
>
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="test"
  />
  <TemplateHeader
    loading={true}
    refresh={[Function]}
    template={
      Object {
        "createdAt": "2020-01-01",
        "defaultFor": Array [],
        "id": "1",
        "name": "test",
        "permissions": Array [],
      }
    }
    topQualifiers={Array []}
  />
  <TemplateDetails
    template={
      Object {
        "createdAt": "2020-01-01",
        "defaultFor": Array [],
        "id": "1",
        "name": "test",
        "permissions": Array [],
      }
    }
  />
  <HoldersList
    groups={Array []}
    onSelectPermission={[Function]}
    onToggleGroup={[Function]}
    onToggleUser={[Function]}
    permissions={
      Array [
        Object {
          "description": "projects_role.user.desc",
          "key": "user",
          "name": "projects_role.user",
        },
        Object {
          "description": "projects_role.codeviewer.desc",
          "key": "codeviewer",
          "name": "projects_role.codeviewer",
        },
        Object {
          "description": "projects_role.issueadmin.desc",
          "key": "issueadmin",
          "name": "projects_role.issueadmin",
        },
        Object {
          "description": "projects_role.securityhotspotadmin.desc",
          "key": "securityhotspotadmin",
          "name": "projects_role.securityhotspotadmin",
        },
        Object {
          "description": "projects_role.admin.desc",
          "key": "admin",
          "name": "projects_role.admin",
        },
        Object {
          "description": "projects_role.scan.desc",
          "key": "scan",
          "name": "projects_role.scan",
        },
      ]
    }
    showPublicProjectsWarning={true}
    users={
      Array [
        Object {
          "login": "<creator>",
          "name": "permission_templates.project_creators",
          "permissions": Array [],
        },
      ]
    }
  >
    <SearchForm
      filter="all"
      onFilter={[Function]}
      onSearch={[Function]}
      query=""
    />
  </HoldersList>
</div>
`;
